core/src/commands/on_runtime_upgrade.rs

```diff
@@ -53,6 +56,87 @@ pub struct Command {
     pub checks: UpgradeCheckSelect,
 }
 
+enum WeightSafety {
+    ProbabySafe,
+    PotentiallyUnsafe,
+}
+
+/// Analyse the given ref_times and return if there is a potential weight safety issue.
+fn analyse_pov(proof: StorageProof, pre_root: H256) -> WeightSafety {
+    let encoded_proof_size = proof.encoded_size();
+    let compact_proof = proof
+        .clone()
+        .into_compact_proof::<sp_runtime::traits::BlakeTwo256>(pre_root)
+        .map_err(|e| {
+            log::error!(target: LOG_TARGET, "failed to generate compact proof: {:?}", e);
+            e
+        })
+        .unwrap_or(CompactProof {
+            encoded_nodes: Default::default(),
+        });
+
+    let compact_proof_size = compact_proof.encoded_size();
+    let compressed_compact_proof = zstd::stream::encode_all(&compact_proof.encode()[..], 0)
+        .map_err(|e| {
+            log::error!(
+                target: LOG_TARGET,
+                "failed to generate compressed proof: {:?}",
+                e
+            );
+            e
+        })
+        .unwrap_or_default();
+
+    let proof_nodes = proof.into_nodes();
+    log::debug!(
+        target: LOG_TARGET,
+        "Proof: 0x{}... / {} nodes",
+        HexDisplay::from(&proof_nodes.iter().flatten().cloned().take(10).collect::<Vec<_>>()),
+        proof_nodes.len()
+    );
+    log::debug!(target: LOG_TARGET, "Encoded proof size: {}", ByteSize(encoded_proof_size as u64));
+    log::debug!(target: LOG_TARGET, "Compact proof size: {}", ByteSize(compact_proof_size as u64),);
+    log::info!(
+        target: LOG_TARGET,
+        "PoV size (zstd-compressed compact proof): {}. For parachains, it's your responsibility \
+        to verify that a PoV of this size fits within any relaychain constraints.",
+        ByteSize(compressed_compact_proof.len() as u64),
+    );
+    if compressed_compact_proof.len() > 4 * 1024 * 1024 {
+        log::warn!(
+            target: LOG_TARGET,
+            "A PoV size of {} is significant. Most relay chains usually accept PoVs up to 5MB. \
+            Proceed with caution.",
+            ByteSize(compressed_compact_proof.len() as u64)
+        );
+        WeightSafety::PotentiallyUnsafe
+    } else {
+        WeightSafety::ProbabySafe
+    }
+}
+
+/// Analyse the given ref_times and return if there is a potential weight safety issue.
+fn analyse_ref_time(ref_time_results: RefTimeInfo) -> WeightSafety {
+    let RefTimeInfo { used, max } = ref_time_results;
+    log::info!(
+        target: LOG_TARGET,
+        "Consumed ref_time: {}s ({:.2}% of max {}s)",
+        used.as_secs_f64(),
+        used.as_secs_f64() / max.as_secs_f64() * 100.0,
+        max.as_secs_f64(),
+    );
+    if used.as_secs_f32() >= max.as_secs_f32() * 0.8 {
```

Comment:
the threshold of 80% seems arbitrary / Polkadot-specific... I'm aware that we must put _something_, but maybe let's extract this constant up and add some comment about it

if you want to be awfully fancy, you can make `WeightSafety` generic over a constant threshold and implement conversion from a fraction :P